// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#pragma once

#include "column/vectorized_fwd.h"
#include "exec/pipeline/operator.h"
#include "exprs/expr.h"
#include "exprs/table_function/table_function_factory.h"
#include "runtime/descriptors.h"
#include "runtime/runtime_state.h"

namespace starrocks::pipeline {
class TableFunctionOperator final : public Operator {
public:
    TableFunctionOperator(OperatorFactory* factory, int32_t id, int32_t plan_node_id, int32_t driver_sequence,
                          const TPlanNode& tnode)
            : Operator(factory, id, "table_function", plan_node_id, driver_sequence), _tnode(tnode) {}

    ~TableFunctionOperator() override = default;

    Status prepare(RuntimeState* state) override;

    void close(RuntimeState* state) override;

    bool has_output() const override;

    bool need_input() const override;

    bool is_finished() const override;

    Status set_finishing(RuntimeState* state) override;

    StatusOr<ChunkPtr> pull_chunk(RuntimeState* state) override;

    Status push_chunk(RuntimeState* state, const ChunkPtr& chunk) override;

    Status reset_state(RuntimeState* state, const std::vector<ChunkPtr>& refill_chunks) override;

private:
    ChunkPtr _build_chunk(const std::vector<ColumnPtr>& output_columns);
    Status _process_table_function(RuntimeState* state);

    const TPlanNode& _tnode;
    const TableFunction* _table_function = nullptr;

    // Slots of output by table function
    std::vector<SlotId> _fn_result_slots;
    // External column slots of the join logic generated by the table function
    std::vector<SlotId> _outer_slots;
    // Slots of table function input parameters
    std::vector<SlotId> _param_slots;

    // Input chunk currently being processed
    ChunkPtr _input_chunk;
    // The current chunk is processed to which row
    size_t _input_chunk_index = 0;
    // The current outer line needs to be repeated several times
    size_t _remain_repeat_times = 0;
    // table function result
    std::pair<Columns, UInt32Column::Ptr> _table_function_result;
    // table function return result end ?
    bool _table_function_result_eos = false;
    // table function param and return offset
    TableFunctionState* _table_function_state = nullptr;

    RuntimeProfile::Counter* _table_function_exec_timer = nullptr;
    RuntimeProfile::Counter* _table_function_exec_counter = nullptr;

    bool _is_finished = false;
};

class TableFunctionOperatorFactory final : public OperatorFactory {
public:
    TableFunctionOperatorFactory(int32_t id, int32_t plan_node_id, const TPlanNode& tnode)
            : OperatorFactory(id, "table_function", plan_node_id), _tnode(tnode) {}

    OperatorPtr create(int32_t degree_of_parallelism, int32_t driver_sequence) override {
        return std::make_shared<TableFunctionOperator>(this, _id, _plan_node_id, driver_sequence, _tnode);
    }

private:
    const TPlanNode& _tnode;
};

} // namespace starrocks::pipeline
